stl/inc/ranges

```diff
@@ -2635,6 +2589,415 @@ namespace ranges {
+        struct _Category_base {};
+
+        // clang-format off
+        template <class _OuterTraits, class _InnerTraits, bool _Deref_is_glvalue>
+            requires _Has_member_iterator_category<_InnerTraits> && _Has_member_iterator_category<_OuterTraits>
+        struct _Category_base<_OuterTraits, _InnerTraits, _Deref_is_glvalue> {
+            using iterator_category =
+                conditional_t<_Deref_is_glvalue
+                           && derived_from<typename _OuterTraits::iterator_category, bidirectional_iterator_tag>
+                           && derived_from<typename _InnerTraits::iterator_category, bidirectional_iterator_tag>,
+                    bidirectional_iterator_tag,
+                    conditional_t<_Deref_is_glvalue
+                               && derived_from<typename _OuterTraits::iterator_category, forward_iterator_tag>
+                               && derived_from<typename _InnerTraits::iterator_category, forward_iterator_tag>,
+                        forward_iterator_tag,
+                        conditional_t<derived_from<typename _OuterTraits::iterator_category, input_iterator_tag>
+                                   && derived_from<typename _InnerTraits::iterator_category, input_iterator_tag>,
+                            input_iterator_tag,
+                            output_iterator_tag>>>;
+        };
+        // clang-format on
+
+        template <bool _Const>
+        class _Iterator : public _Category_base<iterator_traits<iterator_t<_Maybe_const<_Const, _Vw>>>,
+                              iterator_traits<iterator_t<range_reference_t<_Maybe_const<_Const, _Vw>>>>,
+                              is_reference_v<range_reference_t<_Maybe_const<_Const, _Vw>>>> {
+        private:
+            template <bool>
+            friend class _Iterator;
+            template <bool>
+            friend class _Sentinel;
+
+            using _Parent_t  = _Maybe_const<_Const, join_view>;
+            using _Base      = _Maybe_const<_Const, _Vw>;
+            using _SubRng    = range_reference_t<_Base>;
+            using _OuterIter = iterator_t<_Base>;
+            using _InnerIter = iterator_t<_SubRng>;
+
+            // True if and only if the expression *i, where i is an iterator from the outer range, is a glvalue:
+            static constexpr bool _Deref_is_glvalue = is_reference_v<_SubRng>;
+
+            /* [[no_unique_address]] */ _OuterIter _Outer{};
+            /* [[no_unique_address]] */ _InnerIter _Inner{};
+            _Parent_t* _Parent{};
+
+            constexpr void _Satisfy() {
+                auto _Update_inner = [this](_SubRng _Range) -> auto& {
+                    if constexpr (_Deref_is_glvalue) {
+                        (void) this; // Suppress Clang "unused capture" warning
+                        return _Range;
+                    } else {
+                        return (_Parent->_Inner = views::all(_STD move(_Range)));

```

Comment:
Instead of returning an assignment, can we perform this in a separate statement?